src/mbgl/renderer/layers/render_location_component_layer.cpp

```diff
@@ -0,0 +1,895 @@
+#include <array>
+#include <mapbox/cheap_ruler.hpp>
+#include <mbgl/gfx/backend_scope.hpp>
+#include <mbgl/gfx/renderer_backend.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/renderable_resource.hpp>
+#include <mbgl/map/transform_state.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+#include <mbgl/renderer/bucket.hpp>
+#include <mbgl/renderer/layers/render_location_component_layer.hpp>
+#include <mbgl/renderer/paint_parameters.hpp>
+#include <mbgl/style/layers/location_component_layer.hpp>
+#include <mbgl/style/layers/location_component_layer_impl.hpp>
+#include <mbgl/style/layers/location_component_layer_properties.hpp>
+#include <mbgl/util/mat4.hpp>
+
+#include <mapbox/eternal.hpp>
+#include <mbgl/gl/context.hpp>
+#include <mbgl/gl/defines.hpp>
+#include <mbgl/gl/texture.hpp>
+#include <mbgl/gl/texture_resource.hpp>
+#include <mbgl/gl/types.hpp>
+#include <mbgl/platform/gl_functions.hpp>
+#include <mbgl/renderer/image_manager.hpp>
+
+using namespace mbgl::platform;
+namespace mbgl {
+namespace style {
+struct LocationComponentRenderParameters {
+    LocationComponentRenderParameters() = default;
+    LocationComponentRenderParameters(const TransformParameters& tp) : state(&tp.state) {}
+    LocationComponentRenderParameters(const LocationComponentRenderParameters& o) = default;
+    LocationComponentRenderParameters& operator=(const LocationComponentRenderParameters& o) = default;
+
+    double width;
+    double height;
+    double latitude;
+    double longitude;
+    double zoom;
+    double bearing;
+    double pitch;
+    double fieldOfView;
+    std::array<double, 16> projectionMatrix;
+    const TransformState* state = nullptr;
+    ImageManager* imageManager = nullptr;
+    // some testing defaults, for before it gets updated via props
+    double puckBearing = 0.0;
+    LatLng puckPosition = {0, 0};
+    double errorRadiusMeters;
+    mbgl::Color errorRadiusColor{0, 0, 0, 0};
+    mbgl::Color errorRadiusBorderColor{0, 0, 0, 0};
+    int puckSizePx = 0;
+    int puckHatSizePx = 0;
+    int puckShadowSizePx = 0;
+    float puckLayersDisplacement = 0;
+    float perspectiveCompensation = 0;
+    bool useProperProjection = true;
+    std::string puckImagePath;
+    std::string puckShadowImagePath;
+    std::string puckHatImagePath;
+};
+
+// This is a clone of CustomLayerHost. For documentation, refer to that. Leaving it here as it defines a clean interface
+// Not using CustomLayerHost, as it requires a render(CustomLayerRenderParameters).
+class LocationComponentLayerHost {
+public:
+    virtual ~LocationComponentLayerHost() = default;
+    virtual void initialize() = 0;
+    virtual void render(const LocationComponentRenderParameters&) = 0;
+    virtual void contextLost() = 0;
+    virtual void deinitialize() = 0;
+};
+
+class PuckLayerHost : public mbgl::style::LocationComponentLayerHost {
+protected:
+    struct vec2 {
```

Comment:
why not to use Point<float>? math.hpp should have most of aux functions

Reply:
It looks like point class, missing basic vector methods, such as length and normalize(d).
I just tried to use it, and i ended up rewriting half of vec2.